wit/window.wit

```diff
@@ -0,0 +1,228 @@
+        /// A dead key was pressed. The next key event will contain the combined character.
+        dead,
+        /// The key could not be identified.
+        unidentified,
+    }
+    
+    
+    record key-data {
+        modifiers: modifiers,
+        key: key,
+        code: key-code,
+        location: key-location,
+        repeat: bool,
+    }
+    
+    flags mouse-buttons {
+        left,
+        middle,
+        right,
+    }
+    
+    record mouse {
+        position: position,
+        buttons: mouse-buttons,
+    }
+    
+    record scroll {
+        mouse: mouse,
+        x: f32,
+        y: f32,
+    }
+    
+    
+    /// Window system capabilities that are supported.
+    /// Methods that use unsupported capabilities will trap in WASM.
+    flags capabilities {
+        /// The windows have a title that can be set.
+        title,
+        /// The windows have position on a display which can be moved by the program. Not given e.g. on the web, where position is handled by CSS.
+        position,
+        // The windows have icons which can be set by the program.
+        icon,
+        /// Windows can request to go into fullscreen mode. This capability not being present can mean fullscreen windows
+        /// are not supported, or the window is forced fullscreen.
+        fullscreen,
+    }
+    
+    
+    record position {
+        x: u16,
+        y: u16,
+    }
+    
+    record size {
+        width: u16,
+        height: u16,
+    }
+    
+    record create-desc {
+        title: string,
+        position: position,
+        size: size,
+        visible: bool,
+    }
+    
+    /// Checks whether window functionality is available at all in the runtime.
+    available: func() -> bool;
+
+    resource window {
+        constructor(desc: create-desc);
+        
+        connect-graphics-context: func(context: borrow<graphics-context>);
+        
+        set-title: func(title: string);
+        get-title: func() -> string;
+        
+        // TODO add a display interface for querying the displays, so programs can see the resolution and find out where on which screen the window is.
+        set-position: func(position: position);
+        get-position: func() -> position;
+        
+        set-size: func(size: size);
+        get-size: func() -> size;
+        
+        /// Sets the visibility of the window.
+        set-visible: func(visible: bool);
+        
+        subscribe-resize: func() -> pollable;
+        get-resize: func() -> option<size>;
+        
+        subscribe-reposition: func() -> pollable;
+        get-reposition: func() -> option<position>;
+        
+        
+        subscribe-key-down: func() -> pollable;
+        get-key-down: func() -> option<key>;
+        
+        subscribe-key-up: func() -> pollable;
+        get-key-up: func() -> option<key>;
+        
+        
+        subscribe-mouse-down: func() -> pollable;
+        get-mouse-down: func() -> option<mouse>;
```

Comment:
Any reason you're calling these `mouse` instead of `pointer`? I would opt for pointer, like the web, because we wanna cover all kinds of inputs, like touch and pencil.

Reply:
Touch input is a bit more complicated, with multi touch and all, I haven't looked at touch yet. There will be separate events for pointers, but implementations are free to convert e.g. a tap into mouse-down + mouse-up.

Reply:
I'd start with pointer since it covers all the basics (mouse, pen, touch). We can later add in things like multi touch, but I'd prefer if we can start with something broad rather than something narrow.

Reply:
> 我会从指针开始，因为它涵盖了所有基本功能（鼠标、笔、触摸）。我们稍后可以添加多点触摸等功能，但我更希望我们能从广泛的内容开始，而不是从狭窄的内容开始。

strongly agree like https://developer.mozilla.org/zh-CN/docs/Web/API/Pointer_events